cwl/src/main/scala/cwl/preprocessor/CwlPreProcessor.scala

```diff
@@ -0,0 +1,272 @@
+        // passed to this function, as this would indicate that it has already been saladed and pre-processed.
+        case Some(contentRef) if !contentRef.file.equals(reference.file) => original.validParse
+        case _ => flatten.map(_.processedJson)
+      }
+    } yield flattened
+  }
+
+  /**
+    * Convenience method to get the processed workflow as a string.
+    */
+  def preProcessCwlFileToString(file: BFile, cwlRoot: Option[String]): ErrorOr[String] = {
+    preProcessCwlFile(file, cwlRoot).value.unsafeRunSync().map(_.printCompact).toValidated
+  }
+
+  /**
+    * Flatten the cwl reference given already known processed references.
+    */
+  private def flattenCwlReference(cwlReference: CwlReference,
+                                  unProcessedReferences: UnProcessedReferences,
+                                  processedReferences: ProcessedReferences,
+                                  breadCrumbs: Set[CwlReference]): Parse[ProcessedJsonAndDependencies] = {
+    for {
+      // parse the file containing the reference
+      parsed <- saladAndParse(cwlReference.file)
+      // Get a Map[CwlReference, Json] from the parsed file. If the file is a JSON object and only contains one node, the map will only have 1 element 
+      newUnProcessedReferences = mapIdToContent(parsed).toMap
+      // The reference json in the file
+      referenceJson <- newUnProcessedReferences
+        .collectFirst({ case (ref, json) if ref.pointer == cwlReference.pointer => json })
+        .toParse(s"Cannot find a tool or workflow with ID ${cwlReference.fullReference} in file ${cwlReference.file.pathAsString}")
+      // Process the reference json
+      processed <- flattenJson(referenceJson, newUnProcessedReferences ++ unProcessedReferences, processedReferences, breadCrumbs + cwlReference)
+    } yield processed
+  }
+
+  /**
+    * Given a reference from a step's run field, flattens it and return it
+    * @param unProcessedReferences references that have been parsed and saladed (we have the json), but not flattened yet.
+    * @param checkedProcessedReferences references that are fully processed
+    * @param cwlReference reference being processed
+    * @return a new ProcessedReferences Map including this cwlReference processed along with all the dependencies
+    *         that might have been processed recursively.
+    */
+  private def processCwlRunReference(unProcessedReferences: UnProcessedReferences,
+                                     breadCrumbs: Set[CwlReference])
+                                    (checkedProcessedReferences: Parse[ProcessedReferences],
+                                     cwlReference: CwlReference): Parse[ProcessedReferences] = {
+    checkedProcessedReferences flatMap { processedReferences =>
+      // If the reference has already been processed, no need to do anything
+      if (processedReferences.contains(cwlReference)) processedReferences.validParse
```

Comment:
ToL consider pattern matching on a tuple of `(processedReferences.contains(cwlReference), breadCrumbs.contains(cwlReference), unProcessedReferences.get(cwlReference))`